### ISSUE TITLE
Any plan to add support for DoT (preferred) or DoH?

### ISSUE THREAD

**Issue by jol64** [MichaCo/DnsClient.NET#98]
both require a means for authentication in addition to an IPEndpoint. My preference is a server name.
Thanks, Joachim

**Comment by MichaCo** [comment]
I always wanted to add a lot of the security features of the DNS protocol itself. But, mostly because of time constraints, never had the chance to really work on that yet.

Adding the transport layer over HTTP is another thing. 
I have ideas of how that will work, but haven't really played with that yet.

**Comment by Indigo744** [comment]
Hello @MichaCo 

It seems the industry is shifting toward DoH instead of DoT.

**What would be needed to bring DoH (RFC 8484) support to this library?**
DoT could follow of course, but I'm more interested in the former.

I'm willing to take some time to start working on it, but I'm not sure where to begin.
Some pointers would be great.

Thanks.

**Comment by jol64** [comment]
@Indigo744 
I disagree. There is poor adoption in general for either. https://www.dnsfilter.com/blog/dns-over-tls is a good take on why browsers kind of prefer DoH in browsers, but I am convincend organizations would prefer DoT and use existing infrastructure rather that also supports split horizon DNS. And here I believe the biggest challenge is missing support in DHCP for configuration of either DoT or DoH (and yes, I know DHCP is insecure, but it is a reality for mobile clients).
I don´t want you (all) to convince not to implement DoH, but at least it should not preclude adding DoT.
Thanks.

**Comment by Indigo744** [comment]
@jol64 I didn't want to sound negative of course. I've never implied that DoH was better than DoT or anything like that.
Maybe _"shifting toward DoH"_ was too strong. It's just that recent news are more focused on browsers and OS implementation (DoH support in Windows 11 / Server 2022 for instance).

Both exists and both implementation are needed. It's just that we need to pick one and start on it.

**Comment by MichaCo** [comment]
Hi @jol64 and @Indigo744 
I did some testing for DoH manually a while ago. The problem there is that we'd have to rely on HttpClient. The biggest problem I'd have to solve is to manage the lifetime of the HttpClient.

The only implementation for properly managing HttpClients so far is IHttpClientFactory which is in one of those Microsoft extension packages.
I'm not going to include that package as a dependency in DnsClient though, that would be too much in my opinion.

So, I'm looking for a good idea / API approach for that which does not involve any other 3rd part dependency - if you guys have a good idea, let me know

**Comment by jol64** [comment]
@MichaCo would HttpClient itself be ok?
I learned that Cloudflare closes the connection after every single DoT request. I´d consider that a bug, but so far that doesn´t seem a big issue to them and therefore I´d say using the same client only for a single request isn´t a showstopper, no matter whether it is DoT or DoH.
Of course you can also define a transport abstraction interface that can be implemented somewhere else, may be pass that optionally to the DnsClient constructor. Then a heavy user could implement connection pooling etc...

**Comment by Indigo744** [comment]
In our codebase we use [RestClient](https://github.com/NimaAra/Easy.Common/blob/master/Easy.Common/RestClient.cs) from [Easy.Common](https://github.com/NimaAra/Easy.Common). But then again it's another dependency.

We could implement an internal RestClient based on this with only the parts we're interested in. It shouldn't be too hard I guess.

**Comment by MichaCo** [comment]
@Indigo744 After looking at RestClient, that's just a wrapper for HttpClient and doesn't take care of lifetime of the message handler either. Not sure how that would help ;)

**Comment by Indigo744** [comment]
@MichaCo yeah sorry I didn't read properly your message and replied too quickly.

I can think of 2 options:

1. Having a CTOR accepting an optional `HttpClient` in case the user wants to provide an application-wide static one,

1. Implementing using `IHttpClientFactory`, but to avoid forcing this dependency for every users, it is implemented in another project in another namespace, published as another nuget (`DnsClient.DnsOverHttps` for instance),

Thoughts?

**Comment by rmbadmin** [comment]
I think maybe you can let users pass in their own httpclient? My project has its own HttpClientFactory to unify the life cycle of httpclient, maybe you can consider supporting users to pass in their own httpclient?

**Comment by chn-lee-yumi** [comment]
It would be great if it supports DoH. I have some needs about it.

**Comment by jol64** [comment]
@MichaCo I implemented rudimentary DoT support in a local copy, primarily as part of my monitoring application.
Essentially I

- created class NamedIPEndPoint : IPEndPoint that provides the hostname for authentication and server name indication when connecting to the DoT server
- modified one line in LookupClient: `_tcpFallbackHandler = tcpHandler ?? (options.NameServers[0].IPEndPoint is NamedIPEndPoint ? new DnsDotMessageHandler() : new DnsTcpMessageHandler());`
- implemented a rudimentary DnsDotMessageHandler.Query, ignoring async and CancelationToken as I don´t need that in my application, but connecting to a given NamedIPEndPoint, starting TLS, sending request and parsing responses. Most of the last two steps are quick&dirty copy&paste from DnsTcpMessageHandler.QueryInternal(...). I didn´t care about pooling or connection reuse, as afaik that´s not used and I also have no need in my application
- when creating LookupClientOptions I am passing UseTcpOnly=true.

I could share this as a starting point in case you are interested. The redundancy of copy&paste could be reduced by replacing TcpClient in DnsTcpMessageHandler.QueryInternal with a kind of facade to the NetworkStream (my copy uses two args, one NetworkStream, one SslStream/Stream). A similar approach could be used if connection pooling shall be supported. UseTcpOnly could be implicit for NamedIPEndPoint..

**Comment by deepak-nawathe** [comment]
@MichaCo I am using DnsClient.NET and working out quite well, however, now DoT and DoH capabilities are a must. So, any plans or ETA for DoT at least?

**Comment by jol64** [comment]
I succeeded in modifying a fork of DnsClient to support DoT (with arbitrary DoT servers), see comment above, and also to use the Windows Dns Api stub using Vanara.PInvoke.Dns (with the DNS servers configured in the system only, I didn´t get custom servers to work so far, and not all records are supported yet).

The pain points that require forking for any other resolver protocol are
* DnsMessageHandler and other classes required by a DnsMessageHandlers are all internal.
* LookupClient expects two handlers at construction, one udp, one tcp. The tcp-fallback is essentially used for classic DNS only, for DoH there is the opposite occassionally.
* There are insufficient configuration options for any target beyond classic DNS.
* The library obviously obtains some (default) configuration from the OS, but doesn´t understand all of it (my Windows is configured to use DoH only)

What I am kind of thinking of is:
* Custom handlers should compile separately (not excluding that they are merged into the project)
* Configuration options should be handler specific (or extensible at the minimum)
* w.r.t. the UDP-TCP fallback I see two options:
  * move the fallback into a handler that orchestrates the two existing handlers
  * generalize the fallback to allow fallback through a list of arbitrary handlers (that could share/reuse part of configuration) 
* A LookupClient should require one handler only

The reason I am considering to fork or extend DnsClient rather than replacing it is that I like the higher level abstractions compared to other packages.

**Comment by MichaConrad** [comment]
Thanks for the feedback!

FYI

I will have this in the next version of DnsClient. I got a working version for all transport levels at least.
I "just" have to clean it up and finalize some things ;)

This will also be a new major version with some breaking changes..

And yea, a lot of logic will be more abstracted and can be configured / combined as needed. Lookup client will just be a "default" configuration which does the same as LookupClient does today. But you'll be able can create your own client which does some things differently.
